mlflow/langchain/__init__.py

```diff
@@ -648,35 +658,45 @@ def _predict_with_callbacks(
         return results[0] if return_first_element else results
 
     def _prepare_messages(self, data):
-        # numpy array is not json serializable, so we convert it to list
-        # then send it to the model
-        def _convert_ndarray_to_list(data):
-            import numpy as np
-
-            if isinstance(data, np.ndarray):
-                return data.tolist()
-            if isinstance(data, list):
-                return [_convert_ndarray_to_list(d) for d in data]
-            if isinstance(data, dict):
-                return {k: _convert_ndarray_to_list(v) for k, v in data.items()}
-            return data
-
+        """
+        Return a tuple of (preprocessed_data, return_first_element)
+        `preprocessed_data` is always a list,
+        and `return_first_element` means if True, we should return the first element
+        of inference result, otherwise we should return the whole inference result.
+        """
         if isinstance(data, pd.DataFrame):
-            return _convert_ndarray_to_list(data.to_dict(orient="records"))
+            data = data.to_dict(orient="records")
 
         data = _convert_ndarray_to_list(data)
-        if isinstance(self.lc_model, lc_runnables_types()):
-            return [data]
-        if isinstance(data, list) and (
-            all(isinstance(d, str) for d in data) or all(isinstance(d, dict) for d in data)
-        ):
-            return data
+        if not isinstance(data, list):
+            # if the input data is not a list (i.e. single input),
+            # we still need to convert it to a one-element list `[data]`
+            # because `process_api_requests` only accepts list as valid input.
+            # and in this case,
+            # we should return the first element of the inference result
+            # because we change input `data` to `[data]`
+            return [data], True
+        if isinstance(data, list):
+            return data, False
         raise mlflow.MlflowException.invalid_parameter_value(
-            "Input must be a pandas DataFrame or a list of strings "
-            "or a list of dictionaries "
+            "Input must be a pandas DataFrame or a list "
             f"for model {self.lc_model.__class__.__name__}"
         )
 
+    def predict_stream(
+        self,
+        data: Union[Any, Dict[str, Any]],
+        params: Optional[Dict[str, Any]] = None,
+    ) -> List[str]:
+        from mlflow.langchain.api_request_parallel_processor import process_stream_request
+
+        # TODO: Q: shall we put `callback_handlers` in params ?
```

Comment:
I think we should. There is no other way to pass callbacks when `predict_stream` is streamed.

Reply:
It's possible to serialize callback instances when logging a model, but this seems inconvenient.

```
mlflow.langchain.log_model(..., callbacks=...)

# log callbacks in log_model using cloudpickle?
```

Reply:
We can add a method like `_predict_stream_with_callbacks` for now.

Reply:
> We can add a method like `_predict_stream_with_callbacks` for now.

If it's possible to get alignment and expose this in `predict_stream` that'd be optimal to avoid tech debt, but it not, having experimental / private method `_predict_stream_with_callbacks` is fine too

Reply:
@harupy 

> log callbacks in log_model using cloudpickle?

I think we shouldn't freeze `callbacks` when logging model. but we should allow user to use arbitrary `callbacks` after loading the model.

Reply:
The best approach should be making `callbacks` as model inference param, but currently the inference param signature doesn't support complex type like `list[langchain_callback_class]`.

so I'll use the approach of `_predict_stream_with_callbacks` for now.

in future once inference param supports complex type, we can add it as inference param.

Reply:
> in future once inference param supports complex type

I feel like it won't.

Reply:
Idea:

For Validating custom types, we can save a custom validation function when logging model. e.g. `def validate_param(param_value: Any) -> bool` @serena-ruan  Thoughts ?

But we can do it in follow-up PR.

Reply:
I feel like saving a function with the model introduces more complexity as we need to handle ser/de of it, which could be simple but might be a trouble as well. I think we should either expose a `predict_with_callbacks` function at top level, or support skipping validating of certain params, I don't think users really want to validate some complex types like classes

Reply:
+1 this should be (for this PR) exposed via the private method. When the pyfunc implementation is done for general support within pyfunc, this needs to be a public interface argument.